src/server/frontend_wayland/foreign_toplevel_manager_v1.cpp

```diff
@@ -0,0 +1,685 @@
+{
+    surface_stack->add_observer(observer);
+}
+
+mf::ForeignToplevelManagerV1::~ForeignToplevelManagerV1()
+{
+    surface_stack->remove_observer(observer);
+}
+
+void mf::ForeignToplevelManagerV1::stop()
+{
+    send_finished_event();
+    destroy_wayland_object();
+}
+
+// ForeignToplevelHandleV1
+
+mf::ForeignToplevelHandleV1::ForeignToplevelHandleV1(
+    ForeignToplevelManagerV1 const& manager,
+    std::shared_ptr<ms::Surface> const& surface)
+    : mw::ForeignToplevelHandleV1{manager},
+      weak_shell{manager.shell},
+      weak_surface{surface}
+{
+    manager.send_toplevel_event(resource);
+}
+
+void mf::ForeignToplevelHandleV1::send_state(MirWindowFocusState focused, MirWindowState state)
+{
+    switch (state)
+    {
+    case mir_window_state_restored:
+    case mir_window_state_maximized:
+    case mir_window_state_horizmaximized:
+    case mir_window_state_vertmaximized:
+        cached_normal_state = state;
+        cached_fullscreen = false;
+        break;
+
+    case mir_window_state_fullscreen:
+        cached_fullscreen = true;
+        break;
+
+    default:
+        break;
+    }
+
+    wl_array states;
+    wl_array_init(&states);
+
+    if (focused == mir_window_focus_state_focused)
+    {
+        if (uint32_t* state = static_cast<uint32_t*>(wl_array_add(&states, sizeof(uint32_t))))
+            *state = State::activated;
+    }
+
+    switch (state)
+    {
+    case mir_window_state_maximized:
+    case mir_window_state_horizmaximized:
+    case mir_window_state_vertmaximized:
+        if (uint32_t *state = static_cast<uint32_t*>(wl_array_add(&states, sizeof(uint32_t))))
+            *state = State::maximized;
+        break;
+
+    case mir_window_state_fullscreen:
+        if (uint32_t *state = static_cast<uint32_t*>(wl_array_add(&states, sizeof(uint32_t))))
+            *state = State::fullscreen;
+        break;
+
+    case mir_window_state_minimized:
+        if (uint32_t *state = static_cast<uint32_t*>(wl_array_add(&states, sizeof(uint32_t))))
+            *state = State::minimized;
+        break;
+
+    default:
+        break;
+    }
+
+    send_state_event(&states);
+    wl_array_release(&states);
+}
+
+void mf::ForeignToplevelHandleV1::should_close()
+{
+    send_closed_event();
+    weak_surface.reset();
+}
+
+void mf::ForeignToplevelHandleV1::attempt_operation(
+        std::function<void(
+            std::shared_ptr<msh::Shell> const&,
+            std::shared_ptr<ms::Session> const&,
+            std::shared_ptr<ms::Surface> const&)>&& operation)
+{
+    auto const shell = weak_shell.lock();
+    auto const surface = weak_surface.lock();
+    auto const session = surface ? surface->session().lock() : nullptr;
+    if (shell && session && surface)
+    {
+        try
+        {
+            operation(shell, session, surface);
+        }
+        catch (std::out_of_range const&)
+        {
+            log_warning("Attempted operation on invalid surface");
```

Comment:
This catch is no longer necessary. Instead of throwing `std::out_of_range`, `BasicWindowManager` now uses `BasicWindowManager::surface_known()` to print a more useful error and return. We can completely ignore the error case here.